### PR TITLE
[PR-18923] Add coupon validation information

### PR DESCRIPTION
**Pull request by max-si-m**
## Demo

https://deploy-preview-293--talkable-docs.netlify.app/

## Related Stories

[](https://talkable.atlassian.net/browse/PR-18923)

### CODE REVIEW

File: source/advanced_features/coupons.rst [hunk 1]
```diff
@@ -43,3 +43,30 @@ Read :ref:`Create Coupon Webhook Documentation <web_hooks/create_coupon>` for im
...
 
    .. toctree::
+
+Best Practice
+-------------
+
+Coupons should be unique and 3 to 255 characters long.
+All coupons would be converted to uppercase, consider that during code generation, ``fr_coupon`` and ``FR_COUPON`` would be equal.
+
+The is a list of allowed characters:
```
Comment (zhuravel) [inline source/advanced_features/coupons.rst:53]: "The is a" is not a proper English. Let’s rewrite to
```suggestion
Only the following characters are allowed in a coupon code:
```

File: source/advanced_features/coupons.rst [hunk 2]
```diff
@@ -43,3 +43,30 @@ Read :ref:`Create Coupon Webhook Documentation <web_hooks/create_coupon>` for im
...
+   A-Z 0-9 - _ / . % : * + @ & #
+
+.. note::
+   Spaces are forbidden.
+
+Try to avoid ambiguous characters. This simple solution generates a string of easily readable characters for activation codes.
+We do not want people confusing 8 with B, 1 with I, 0 with O, L with 1, etc.
+
+
+.. code-block:: ruby
```
Comment (zhuravel) [inline source/advanced_features/coupons.rst:66]:
```suggestion

.. code-block:: ruby
```

File: source/advanced_features/coupons.rst [hunk 3]
```diff
@@ -43,3 +43,30 @@ Read :ref:`Create Coupon Webhook Documentation <web_hooks/create_coupon>` for im
...
+
+   # Generates a random string from a set of easily readable characters
+   def generate_activation_code(size = 6)
+     charset = %w{ 2 3 4 6 7 9 A C D E F G H J K M N P Q R T V W X Y Z}
+     (0...size).map { charset.to_a[rand(charset.size)] }.join
```
Comment (zhuravel) [inline source/advanced_features/coupons.rst:71]: There is a simpler way to achieve that. Not sure why other code is so complex.
```suggestion
     "234679ACDEFGHJKMNPQRTVWXYZ".chars.sample(size).join
```

File: source/advanced_features/coupons.rst [hunk 4]
```diff
@@ -28,8 +28,8 @@ In this case each, customer will receive unique coupon code from the pool availa
 
 There are several steps that need to be taken for this scenario:
 
-1. Create a coupon list at "Manage Coupon Lists" page
-2. Generate a list of coupons on the merchant site using mass coupon generation tool
+1. Create a coupon list at ``All reports → Coupon Lists`` page
+2. Generate a list of coupons on the merchant site using mass coupon generation tool. Check the Best Practice section, before generating coupons.
```
Comment (zhuravel) [inline source/advanced_features/coupons.rst:32]: Add a link to the section.
```suggestion
2. Generate a list of coupons on the merchant site using mass coupon generation tool. Check the `Best Practice`_ section before generating coupons.
```

File: source/advanced_features/coupons.rst (hunk 3 above)
Comment (zhuravel) [inline source/advanced_features/coupons.rst:71]: `to_a` seems redundant.
```suggestion
     (0...size).map { charset[rand(charset.size)] }.join
```

File: source/advanced_features/coupons.rst [hunk 5]
```diff
@@ -43,3 +43,30 @@ Read :ref:`Create Coupon Webhook Documentation <web_hooks/create_coupon>` for im
...
+
+Try to avoid ambiguous characters. This simple solution generates a string of easily readable characters for activation codes.
+We do not want people confusing 8 with B, 1 with I, 0 with O, L with 1, etc.
+
+
+.. code-block:: ruby
+
+   # Generates a random string from a set of easily readable characters
+   def generate_activation_code(size = 6)
+     charset = %w{ 2 3 4 6 7 9 A C D E F G H J K M N P Q R T V W X Y Z}
```
Comment (zhuravel) [inline source/advanced_features/coupons.rst:70]:
```suggestion
     charset = "234679ACDEFGHJKMNPQRTVWXYZ".chars
```

File: source/advanced_features/coupons.rst [hunk 6]
```diff
@@ -43,3 +43,30 @@ Read :ref:`Create Coupon Webhook Documentation <web_hooks/create_coupon>` for im
 .. container:: hidden
 
    .. toctree::
+
+Best Practice
```
Comment (dgtv) [inline source/advanced_features/coupons.rst:47]: Actually it is not "Best Practice" but requirements, so lets rename this section accordingly. The only "best practice" here is "avoid ambiguous characters" note

File: source/advanced_features/coupons.rst (hunk 5 above)
Comment (dgtv) [reply source/advanced_features/coupons.rst:70]: I'd have here as little as possibly "Ruby sugar" to have it easy-to-read by people from other languages, so lets use general constructions like array of chars:
```ruby
charset = ["2", "3", "4", "6", "7", "9", "A", "C", "D", "E", "F", "G", "H", "J", "K", "M", "N", "P", "Q", "R", "T", "V", "W", "X", "Y", "Z"]
```

**Review by zhuravel — APPROVED** [review]

File: source/advanced_features/coupons.rst [hunk 7]
```diff
@@ -43,3 +43,29 @@ Read :ref:`Create Coupon Webhook Documentation <web_hooks/create_coupon>` for im
...
+Only the following characters are allowed in a coupon code:
+
+.. code-block:: text
+
+   A-Z 0-9 - _ / . % : * + @ & #
+
+.. note::
+   Spaces are forbidden.
+
+Try to avoid ambiguous characters. This simple solution generates a string of easily readable characters for activation codes.
```
Comment (dgtv) [inline source/advanced_features/coupons.rst:62]: I'd move this to `Note` section as it is not requirement but a notice

**Review by dgtv — APPROVED** [review]

**Pull request merged**